### PR TITLE
use unique_ptr in src/server.h

### PR DESCRIPTION
**Pull request by zz-jason**
Signed-off-by: Jian Zhang <zjsariel@gmail.com>

to #663

### CODE REVIEW

**Comment by tisonkun** [comment]
This patch can be blocked by #645 as `missingInclude` is tricky to address. See also https://github.com/apache/incubator-kvrocks/pull/645#issuecomment-1162042973.

**Review by tisonkun — CHANGES_REQUESTED** [review]
@zz-jason There're other failure in CI tasks. Please make sure that with you patch you can still compile kvrocks and pass `./cppcheck.sh` at least.

File: src/server.h [hunk 1]
```diff
@@ -198,10 +198,10 @@ class Server {
 
   Stats stats_;
   Engine::Storage *storage_;
-  Cluster *cluster_;
+  std::unique_ptr<Cluster> cluster_;
   static std::atomic<int> unix_time_;
-  class SlotMigrate *slot_migrate_ = nullptr;
-  class SlotImport *slot_import_ = nullptr;
+  std::unique_ptr<class SlotMigrate> slot_migrate_;
+  class SlotImport *slot_import_;
```
Comment (PragmaTwice) [inline src/server.h]:
```suggestion
  class SlotImport *slot_import_ = nullptr;
```
fix the cppcheck warning in CI

**Comment by PragmaTwice** [comment]
Could you remove `--suppress=noCopyConstructor:src/server.cc --suppress=noOperatorEq:src/server.cc` in `./cppcheck.sh` to make CI pass? Sorry for discommodity.

**Comment by zz-jason** [comment]
I find the help message of `cppcheck` is:

```
--suppress=<spec>    Suppress warnings that match <spec>. The format of
                     <spec> is:
                     [error id]:[filename]:[line]
                     The [filename] and [line] are optional. If [error id]
                     is a wildcard '*', all error ids match.
```

Does `--suppress=noCopyConstructor:src/server.cc` mean to ignore the `noCopyConstructor` warning for `src/server.cc`?

**Comment by PragmaTwice** [comment]
> I find the help message of `cppcheck` is:
> 
> ```
> --suppress=<spec>    Suppress warnings that match <spec>. The format of
>                      <spec> is:
>                      [error id]:[filename]:[line]
>                      The [filename] and [line] are optional. If [error id]
>                      is a wildcard '*', all error ids match.
> ```
> 
> Does `--suppress=noCopyConstructor:src/server.cc` mean to ignore the `noCopyConstructor` warning for `src/server.cc`?

Yeah, my guess is that the `noCopyConstructor` warning no longer appears because you changed the type of some of `Server`'s fields to `unique_ptr` (which has no copy ctor) so that `Server` can no longer copy constructible. 
At this point cppcheck reports `unmatchedSuppression` because we added an useless `--suppress` (after this change).
We will consider using clang tidy instead of cppcheck in the future.
```
src/server.cc:-1:0: information: Unmatched suppression: noCopyConstructor [unmatchedSuppression]
^
src/server.cc:-1:0: information: Unmatched suppression: noOperatorEq [unmatchedSuppression]
^
```
Actually you can just add `--suppress=unmatchedSuppression` instead of remove anything.

**Review by PragmaTwice — APPROVED** [review]
LGTM. cc @tisonkun

**Review by tisonkun — APPROVED** [review]
LGTM.

@git-hulk @PragmaTwice it seems we don't have `delete cluster_` to eliminate in this patch, did we leak memory previously on these field?

**Comment by PragmaTwice** [comment]
> LGTM.
> 
> @git-hulk @PragmaTwice it seems we don't have `delete cluster_` to eliminate in this patch, did we leak memory previously on these field?

I also wonder that. 🤣

**Comment by git-hulk** [comment]
@tisonkun @PragmaTwice Yes, we do leak this field even it wouldn't cause problem.

**Review by git-hulk — APPROVED** [review]

**Comment by git-hulk** [comment]
Thanks for @zz-jason contribution, merging...

**Pull request merged**